// hdd-space/index.js
"use strict";
var child_process_1 = require('child_process');
var outputParser_1 = require('./outputParser');
;
;
function getTotal(parts) {
    return parts.reduce(function (total, part) {
        total.size += part.size;
        total.free += part.free;
        return total;
    }, {
        size: 0,
        free: 0
    });
}
;
function parseDf(output) {
    return outputParser_1.parseOutput({
        size: {
            name: '1024-blocks',
            align: 'right',
            re: /\d+/
        },
        mp: {
            name: 'Mounted on',
            align: 'left',
            re: /[\s\S]*/
        },
        free: {
            name: 'Available',
            align: 'right',
            re: /\d+/
        }
    }, output);
}
;
function getUnixInfo(callback) {
    /*
        [df] output example:

        Filesystem     1K-blocks    Used Available Use% Mounted on
        rootfs          15506408 1149728  13708188   8% /
        /dev/root       15506408 1149728  13708188   8% /
        devtmpfs          447624       0    447624   0% /dev
        tmpfs              89548     208     89340   1% /run
        tmpfs               5120       0      5120   0% /run/lock
        tmpfs             179080       0    179080   0% /run/shm

    */
    child_process_1.exec('df -Pk', function (err, output) {
        var parsed = parseDf(output.trim());
        var root = null;
        var parts = parsed
            .map(function (part) {
            var res = {
                size: parseInt(part.size, 10) * 1024,
                free: parseInt(part.free, 10) * 1024,
                place: part.mp,
                mountOn: part.mp
            };
            if (part.mp === "/") {
                root = res;
            }
            ;
            return res;
        });
        var resultObj = {
            parts: parts,
            total: root
        };
        callback(resultObj);
    });
}
;
function isNaN(val) {
    return val !== val;
}
function getWindowsInfo(callback) {
    /*
        [wmic logicaldisk get size,freespace,caption] output example:

        Caption  FreeSpace    Size
        C:       1286164480   34359734272
        D:       1864638464   50925137920
        E:
        F:       77553082368  990202818560
        G:
        L:


    */
    child_process_1.exec('wmic logicaldisk get size,freespace,caption', function (err, output) {
        var parts = output
            .split('\n')
            .slice(1) // remove header
            .map(function (line) {
            var lineParts = line.split(/[\s]+/g);
            var partInfo = {
                place: lineParts[0],
                letter: lineParts[0],
                free: parseInt(lineParts[1]),
                size: parseInt(lineParts[2])
            };
            if (isNaN(partInfo.free) ||
                isNaN(partInfo.size) ||
                partInfo.place == '') {
                return null;
            }
            ;
            return partInfo;
        })
            .filter(function (part) {
            return !!part;
        });
        var resultObj = {
            parts: parts,
            total: getTotal(parts)
        };
        callback(resultObj);
    });
}
;
var formaters = {
    "byte": function (size) { return (size) + ' Bytes'; },
    "bit": function (size) { return (size * 8) + ' Bits'; },
    "kb": function (size) { return (size / 1024).toFixed(2) + ' KB'; },
    "mb": function (size) { return (size / 1024 / 1024).toFixed(2) + ' MB'; },
    "gb": function (size) { return (size / 1024 / 1024 / 1024).toFixed(2) + ' GB'; },
    "tb": function (size) { return (size / 1024 / 1024 / 1024 / 1024).toFixed(2) + ' TB'; },
    "pb": function (size) { return (size / 1024 / 1024 / 1024 / 1024 / 1024).toFixed(2) + ' PB'; },
    "auto": function (size) {
        if (size === 0) {
            return formaters.byte(size);
        }
        ;
        var scale = Math.floor(Math.log(size) / Math.log(1024));
        switch (scale) {
            case 0: return formaters.byte(size);
            case 1: return formaters.kb(size);
            case 2: return formaters.mb(size);
            case 3: return formaters.gb(size);
            case 4: return formaters.tb(size);
            default: return formaters.pb(size);
        }
    }
};
function formatResult(opts, res) {
    var formater = opts.format;
    if (typeof formater !== "function") {
        formater = formaters[formater.toLowerCase()];
    }
    ;
    var newRes = {
        parts: res.parts.map(function (part) {
            var newPart = {
                free: formater(part.free),
                size: formater(part.size),
                place: part.place
            };
            if ('letter' in part) {
                newPart.letter = part.letter;
            }
            ;
            if ('mountOn' in part) {
                newPart.mountOn = part.mountOn;
            }
            ;
            return newPart;
        }),
        total: {
            free: formater(res.total.free),
            size: formater(res.total.size)
        }
    };
    return newRes;
}
;
function getCrossPlatformInfo(opts, callback) {
    if (arguments.length < 2) {
        callback = opts;
        opts = { format: function (size) { return size; } };
    }
    ;
    var func = process.platform === "win32"
        ? getWindowsInfo
        : getUnixInfo;
    func(function (res) {
        callback(formatResult(opts, res));
    });
}
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = getCrossPlatformInfo;
;
module.exports = getCrossPlatformInfo;
